### ISSUE TITLE
Impossible to compile with PyInstaller or Py2Exe

### ISSUE THREAD

**Issue by RazenIW** [kmaork/pyinjector#17]
Hello,

I have been using your library and this is great work, but I am having an issue when trying to run my program once it's compiled to a Windows executable.

It compiles without errors with PyInstaller but this is what I get when I run it : 

```
C:\Users\razen\PycharmProjects\RAT\dist>main.exe
Traceback (most recent call last):
  File "main.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "pyinjector\__init__.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "pyinjector\pyinjector.py", line 6, in <module>
AttributeError: 'NoneType' object has no attribute 'origin'
[12032] Failed to execute script 'main' due to unhandled exception!
```

This is what I get when running the executable from Py2Exe : 

```
C:\Users\razen\PycharmProjects\RAT\dist>main.exe
Traceback (most recent call last):
  File "main.py", line 1, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "pyinjector\__init__.pyc", line 1, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "pyinjector\pyinjector.pyc", line 6, in <module>
  File "importlib\util.pyc", line 90, in find_spec
  File "importlib\util.pyc", line 32, in resolve_name
ImportError: no package specified for '.libinjector' (required for relative module names)
```

This one goes further and it seems like it is an import issue.

I read somewhere that those compilers can only import libraries using the "import" command.

I also see that you import ".libinjector" another way :

`libinjector_path = find_spec('.libinjector', __package__).origin`

And this is probably what's causing the issue. 

Is there any possible fix ?

Thanks !

**Comment by madaraXor** [comment]
Hello, I had the same problem, I solved it by replacing the line:
`libinjector_path = find_spec('.libinjector', __package__).origin`
from pyinjector.py
by :
`libinjector_path = "libinjector.cp37-win_amd64.pyd"`

Then you have to add the file "libinjector.cp37-win_amd64.pyd"
to the executable, I don't know how to do with pyinstaller or py2exe, I use nuitka, here is the argument to add to the nuitka command :
--include-data-file=C:\Users\\********\AppData\Local\Programs\Python\Python37\lib\site-packages\pyinjector\libinjector.cp37-win_amd64.pyd=libinjector. cp37-win_amd64.pyd

hoping to help

J'avais pas vu que t'étais français, si ta besoin d'aide pour faire fonctionner hésite pas

**Comment by h8rr** [comment]
Happening to me :/

Tried this, to no avail

> Hello, I had the same problem, I solved it by replacing the line: `libinjector_path = find_spec('.libinjector', __package__).origin` from pyinjector.py by : `libinjector_path = "libinjector.cp37-win_amd64.pyd"`

> Then you have to add the file "libinjector.cp37-win_amd64.pyd" to the executable, I don't know how to do with pyinstaller or py2exe, I use nuitka, here is the argument to add to the nuitka command : --include-data-file=C:\Users\********\AppData\Local\Programs\Python\Python37\lib\site-packages\pyinjector\libinjector.cp37-win_amd64.pyd=libinjector. cp37-win_amd64.pyd

**Comment by h8rr** [comment]
Error:

`Traceback (most recent call last):
  File "PyInstaller\loader\pyimod03_ctypes.py", line 53, in __init__
  File "ctypes\__init__.py", line 376, in __init__
FileNotFoundError: Could not find module 'C:\Users\monte\AppData\Local\Temp\_MEI239162\libinjector.cp37-win_amd64.pyd' (or one of its dependencies). Try using the full path with constructor syntax.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "compile.pyw", line 14, in <module>
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
  File "modules\bacon.py", line 14, in <module>
  File "PyInstaller\loader\pyimod03_ctypes.py", line 55, in __init__
pyimod03_ctypes.install.<locals>.PyInstallerImportError: Failed to load dynlib/dll 'C:\\Users\\monte\\AppData\\Local\\Temp\\_MEI239162\\libinjector.cp37-win_amd64.pyd'. Most likely this dynlib/dll was not found when the application was frozen`

**Comment by TerraFaster** [comment]
You can do this:
1. Copy the `libinjector.cpPYTHON_VERSION-win_OS_BIT.pyd` file to your project data folder.
• `PYTHON_VERSION` is your python version, in my case i have `Python 3.10` so `PYTHON_VERSION` is `310`.
• `OS_BIT` you can get using `import platform; platform.machine().lower()`, in my case it is `amd64`.
• In my case, filename is `libinjector.cp310-win_amd64.pyd`.

2. Now you need to patch pyinjector's `api.py`:
Replace line `libinjector_path = find_spec('.libinjector', __package__).origin` with this:
```
import sys

if getattr(sys, 'frozen', False):
    file_name = f'libinjector.cpPYTHON_VERSION-win_OS_BIT.pyd'
    libinjector_path = f'{sys._MEIPASS}/YOUR_PROJECT_DATA_FOLDER/{file_name}'

else:
    libinjector_path = find_spec('.libinjector', __package__).origin
```
• You need to replace `YOUR_PROJECT_DATA_FOLDER` with the name of the folder to where you copied file in `Step 1`.
• Also you need to replace `PYTHON_VERSION` and `OS_BIT` to your values.

3. Now you can package your python application using PyInstaller, don't forget to add `--add-data YOUR_PROJECT_DATA_FOLDER` argument to your PyInstaller CLI command.

**Comment by kmaork** [comment]
Thanks everyone for the issue and the notes. As far as I understand it, this issue is due to me being too lazy to write an actual python extension, and interfacing with native code using ctypes instead. A good workaround is indeed to include the generated `pyd`, but to solve this properly, I finally opened [a pr](https://github.com/kmaork/pyinjector/pull/29) to use python c api. It seems to work for me with py2exe. Does it work for you people with nuitka and pyinstaller?

**Comment by Whitesun123** [comment]
> Thanks everyone for the issue and the notes. As far as I understand it, this issue is due to me being too lazy to write an actual python extension, and interfacing with native code using ctypes instead. A good workaround is indeed to include the generated `pyd`, but to solve this properly, I finally opened [a pr](https://github.com/kmaork/pyinjector/pull/29) to use python c api. It seems to work for me with py2exe. Does it work for you people with nuitka and pyinstaller?

Nuitka only works if I don't use any options like --standalone and --onefile (nuitka main.py)
For pyinstaller I use auto-py-to-exe because of the GUI and the app just returns an error when trying to run it.

**Comment by kmaork** [comment]
Thanks for checking @Whitesun123. So, you checked out the branch I posted, installed it and these errors still occurred? Can you please post your code/commands and the errors you got with nuitka and pyinstaller?

**Comment by Whitesun123** [comment]
I’m on my phone so I can only post the Pyinstaller error:

```
Traceback (most recent call last):
  File "main.py", line 11, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
  File "pyinjector\__init__.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
  File "pyinjector\api.py", line 7, in <module>
AttributeError: 'NoneType' object has no attribute 'origin'
```
in the code all I do is “inject(pid, dll)”

**Comment by kmaork** [comment]
This error would only occur in the current version of pyinjector, because the code in the stack trace you posted does not exist in the feature brach. Can you make sure that pyinjector from the feature branch is installed before running pyinstaller?

**Comment by Whitesun123** [comment]
> This error would only occur in the current version of pyinjector, because the code in the stack trace you posted does not exist in the feature brach. Can you make sure that pyinjector from the feature branch is installed before running pyinstaller?

Sorry for the late response, switching to the 'python_c_api' branch seems to fix the issue with nuitka and pyinstaller. Both of them built a runnable binary file. Thank you for your help.

Nuitka command:
`nuitka --windows-icon-from-ico=C:\Users\white\Desktop\Code\Python\dll_injector\titleLogo.ico --standalone --onefile --disable-console --include-data-file=titleLogo.png=. main.py`

Pyinstaller command:
`pyinstaller --noconfirm --onefile --windowed --icon "C:/Users/white/Desktop/Code/Python/dll_injector/titleLogo.ico" --add-data "C:/Users/white/Desktop/Code/Python/dll_injector/titleLogo.png;."  "C:/Users/white/Desktop/Code/Python/dll_injector/main.py"`

**Comment by kmaork** [comment]
Thank you for your help :) I released `pyinjector==1.3.0` in which this issue is fixed.
